File: ocaml/lib/code_analysis/syntax/node/expression/parenthesized.ml
open Types

type t = parenthesized

let make open_parenthesis_token expression close_parenthesis_token =
  {
    parenthesized_open_parenthesis_token = open_parenthesis_token;
    parenthesized_expression = expression;
    parenthesized_close_parenthesis_token = close_parenthesis_token;
  }

let open_parenthesis_token x = x.parenthesized_open_parenthesis_token
let expression x = x.parenthesized_expression
let close_parenthesis_token x = x.parenthesized_close_parenthesis_token

File: hs/src/Minsk/ExpressionSyntax/Parenthesized.hs
module Minsk.ExpressionSyntax.Parenthesized
  ( ParenthesizedExpressionSyntax (..),
    parenthesizedExpressionSyntax,
  )
where

import BasicPrelude
import Minsk.ExpressionSyntax (ExpressionSyntax (ExpressionSyntax), IsExpressionSyntax)
import qualified Minsk.SyntaxKind as SyntaxKind
import Minsk.SyntaxNode (IsSyntaxNode (..), SyntaxNode (SyntaxNode), node)
import Minsk.SyntaxToken (SyntaxToken)

data ParenthesizedExpressionSyntax = ParenthesizedExpressionSyntax
  { openParenthesisToken :: SyntaxToken,
    expression :: ExpressionSyntax,
    closeParenthesisToken :: SyntaxToken
  }

parenthesizedExpressionSyntax :: SyntaxToken -> ExpressionSyntax -> SyntaxToken -> ExpressionSyntax
parenthesizedExpressionSyntax open e close = ExpressionSyntax $ ParenthesizedExpressionSyntax open e close

instance IsSyntaxNode ParenthesizedExpressionSyntax where
  node (ParenthesizedExpressionSyntax open e close) =
    SyntaxNode SyntaxKind.ParenthesizedExpression [node open, node e, node close] mempty

instance IsExpressionSyntax ParenthesizedExpressionSyntax
